File: 1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.py
class Solution:
    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
        graph = [[int(1e9) for i in range(n)] for j in range(n)]
        result = defaultdict(list)

        for f, t, w in edges:
            graph[f][t] = w
            graph[t][f] = w

        for i in range(n):
            graph[i][i] = 0

        for i in range(n):
            for j in range(n):
                for k in range(n):
                    if graph[j][k] <= graph[j][i] + graph[i][k]: continue
                    graph[j][k] = graph[j][i] + graph[i][k]

        for i in range(n):
            cnt = 0
            for j in range(n):
                if i != j and graph[i][j] <= distanceThreshold:
                    cnt += 1
            result[cnt].append(i)
        
        return sorted(sorted(result.items(), key=lambda x: x[0])[0][1])[-1]

File: 1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.java
class Solution {
    public int findTheCity(int n, int[][] edges, int distanceThreshold) {
        int[][] graph = new int[n][n];
        Map<Integer, List<Integer>> dist = new HashMap<>();

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i == j) continue;
                graph[i][j] = (int) Math.pow(10, 9);
            }
        }

        for (int i = 0; i < edges.length; i++) {
            graph[edges[i][0]][edges[i][1]] = edges[i][2];
            graph[edges[i][1]][edges[i][0]] = edges[i][2];
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < n; k++) {
                    if (graph[j][k] <= graph[j][i] + graph[i][k]) continue;
                    graph[j][k] = graph[j][i] + graph[i][k];
                }
            }
        }

        for (int i = 0; i < n; i++) {
            int cnt = 0;
            for (int j = 0; j < n; j++) {
                if (i != j && graph[i][j] <= distanceThreshold) cnt++;
            }
            if (!dist.containsKey(cnt)) dist.put(cnt, new ArrayList<>());
            dist.get(cnt).add(i);
        }

        List<Integer> keySet = new ArrayList<>(dist.keySet());
        Collections.sort(keySet);

        List<Integer> result = dist.get(keySet.get(0));
        Collections.reverse(result);

        return result.get(0);
    }
}
